### PR TITLE
feat: add vercel-build action input for local Vercel build step

### PR DESCRIPTION
**Pull request by amondnet**
## Track

- Track: `auto-vercel-build-20260430`
- Spec: `.please/docs/tracks/active/auto-vercel-build-20260430/spec.md`
- Plan: `.please/docs/tracks/active/auto-vercel-build-20260430/plan.md`

Refs #360

## Summary

Add an opt-in `vercel-build` action input that, when set to `true`, executes the official Vercel CLI workflow locally before deployment: `vercel pull --environment=<target>` → `vercel build` → upload `.vercel/output` via the existing prebuilt deploy path.

This mirrors the [Vercel KB recommended GitHub Actions workflow](https://vercel.com/kb/guide/how-can-i-use-github-actions-with-vercel#configuring-github-actions-for-vercel) and lets users build inside CI (with their own secrets and runtime) instead of relying on Vercel's remote build.

## Behavior

| `vercel-build` | `prebuilt` | Behavior |
|---|---|---|
| `false` (default) | any | **Unchanged** — current source-upload behavior |
| `true` | `false` | Run `vercel pull` + `vercel build` locally, then deploy `.vercel/output` as prebuilt |
| `true` | `true` | Fail fast — mutually exclusive |

## Status

- [x] Spec & plan
- [ ] Implementation (16 tasks, in progress)
- [ ] Tests (>80% coverage)
- [ ] Integration test (emulate.dev)
- [ ] README update
- [ ] dist/ rebuild

---
## Summary by cubic
Adds an opt-in `vercel-build` input to run `vercel pull` and `vercel build` inside the action, then deploy `.vercel/output` or a custom `vercel-output-dir` via the prebuilt path. Adds failure comments for broken builds, keeps current behavior by default, and hardens token handling. Addresses #360.

- **New Features**
  - `vercel-build` (default: `false`): runs `vercel pull --environment=<target>` → `vercel build` (`--prod` for production; respects `--output <dir>` when `vercel-output-dir` is set), then uploads the build output; on failure, exits non-zero and posts a short PR/commit comment (when enabled).
  - Input guards: mutually exclusive with `prebuilt` and with a non-empty `vercel-args`; both fail fast with a clear error at config parse time.
  - Forwards `build-env` to `vercel build` and passes `--scope` when a team scope is set.

- **Bug Fixes**
  - Hardened logging and comments: pass the Vercel token via `VERCEL_TOKEN` with `@actions/exec` `silent: true` to avoid token leakage; escape triple backticks in failure comments to prevent Markdown injection.
  - Bound captured build stdout/stderr to ~64 KB per stream to prevent runner OOM on long builds while keeping the tail in failure comments.
  - Tests: clear `GITHUB_ACTIONS` in the Vitest unit project to prevent auto-invoking `run()` on import.
  - Rebuilt `dist/` to match source; no behavior change.

<sup>Written for commit faeeff0455a6f0a6fb08cc1029221961dd7bf616. Summary will update on new commits. <a href="https://cubic.dev/pr/amondnet/vercel-action/pull/361?utm_source=github">Review in cubic</a></sup>

## Verification Checklist

Manual verification (run after T013):

- [ ] **Happy path (preview)**: In `example/nextjs`, set `vercel-build: 'true'`, `prebuilt: 'false'`. Run the action against a real Vercel project (preview token). Expect: log shows `vercel pull` then `vercel build` invocations, `.vercel/output` is uploaded, deployment URL returned.
- [ ] **Happy path (production)**: Set `target: production` + `vercel-build: 'true'`. Expect: `vercel pull --environment=production` and `vercel build --prod` invoked.
- [ ] **Mutual exclusivity**: Set `vercel-build: 'true'` AND `prebuilt: 'true'`. Expect: action fails immediately with a clear conflict message; no API calls made.
- [ ] **Custom output dir**: Set `vercel-build: 'true'` + `vercel-output-dir: 'custom/out'`. Expect: `vercel build --output <abs>/custom/out` invoked, deploy reads from same path.
- [ ] **Build failure**: In a fixture project, introduce a syntax error so `vercel build` fails. Expect: action exits non-zero, GitHub Actions log shows the build error, PR comment is posted with truncated tail (when `github-comment: true`).
- [ ] **`github-comment: false` suppression**: With `vercel-build: 'true'` and `github-comment: 'false'`, trigger a build failure. Expect: action fails non-zero, **no** PR/commit comment posted.
- [ ] **Backward compatibility**: With `vercel-build` unset (or `false`), run an existing-style deployment. Expect: identical behavior to current main branch — no `vercel pull`/`vercel build` invocations.
- [ ] **Token redaction**: Inspect logs from steps above. The Vercel token MUST NOT appear in any log line, comment, or error message. Token is now passed via `VERCEL_TOKEN` env var (never via argv).

Automated:

- [x] `pnpm test` — all 285 tests passing
- [x] `pnpm test --coverage` — `src/vercel-build.ts` 94%+ coverage
- [x] `pnpm run lint` — 0 errors
- [x] `pnpm run build` — clean dist/ rebuild

### CODE REVIEW

File: .please/docs/tracks/active/auto-vercel-build-20260430/spec.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+The current default behavior (uploading source via `@vercel/client` for remote build) is preserved when `vercel-build` is `false` (the default), maintaining backward compatibility with existing pipelines.
+
+## Requirements
+
+### Functional Requirements
+
+- [ ] FR-1: Add a new boolean action input `vercel-build` to `action.yml` (default: `false`).
+- [ ] FR-2: When `vercel-build: true` and `prebuilt: false`, the action MUST execute, in order, in the configured `working-directory`:
+  1. `vercel pull --yes --environment=<target> --token=<vercel-token>` (where `<target>` is `production` or `preview` based on the existing `production`/`target` inputs)
+  2. `vercel build [--prod] --token=<vercel-token>` (the `--prod` flag is included only when targeting production)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/tracks/active/auto-vercel-build-20260430/spec.md]: 💡 Implemented with alternative approach ([f7a132f](https://github.com/amondnet/vercel-action/commit/f7a132f3e4760d11dcdf3cab0e3113e43d907a0f))

- **Suggestion**: pass token via env var; include `[--scope <scope>]` in spec text
- **Applied**:
  - Token transport via env var was already fixed in commit `6cf21ad` (predates this review). `runVercelPull` and `runVercelBuild` set `VERCEL_TOKEN` in the exec env and pass `silent: true` to suppress the `[command]…` echo. The CLI argv no longer contains `-t` / `--token`.
  - `--scope` was already implemented in code via `appendScope()` (`src/vercel-build.ts:88-93`). I've now updated the spec text (FR-2) to use the documented `[--scope <scope>]` notation and added an explicit note that auth flows through `VERCEL_TOKEN`.
- **Reason**: alternative because the runtime fix predated your review; spec text update brings the docs in line with the existing implementation.

Thanks for the catch, @gemini-code-assist!

File: .please/docs/tracks/active/auto-vercel-build-20260430/spec.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  3. Treat the resulting `.vercel/output` directory as the deploy artifact, using the existing prebuilt deploy code path (`config.prebuilt = true`, `vercelOutputDir` defaulted to `<working-directory>/.vercel/output`).
```
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/tracks/active/auto-vercel-build-20260430/spec.md]: ✅ Applied in [f7a132f](https://github.com/amondnet/vercel-action/commit/f7a132f3e4760d11dcdf3cab0e3113e43d907a0f)

- **Files**:
  - `src/vercel-build.ts` — `runVercelBuild` now passes `--output <dir>` whenever `config.vercelOutputDir` is set; `runBuildStep` returns the user's resolved `vercelOutputDir` (relative paths resolved against `working-directory`) so the deploy reads from the same path the build wrote to.
  - `spec.md` — FR-2 step 2 now shows `[--output <dir>]` and a new AC-7 covers the contract.
  - 5 new unit tests cover the omission, absolute-path, and relative-resolution cases.

This was a real bug: previously a user combining `vercel-build: true` with a custom `vercel-output-dir` would have the build silently write to the default `.vercel/output` while the deploy looked at the configured path. Great catch — thanks @gemini-code-assist!

File: .please/docs/tracks/active/auto-vercel-build-20260430/plan.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+```
+run() in src/index.ts
+  │
+  ├─► getActionConfig()                    [config.ts]
+  │     └─ validates vercelBuild XOR prebuilt   ← FR-3
+  │
+  ├─► [if config.vercelBuild]
+  │     └─► runVercelBuild(config)          [NEW: vercel-build.ts]
+  │           ├─ exec: npx vercel pull --yes --environment=<target>
+  │           └─ exec: npx vercel build [--prod]
```
[the remark this answers is not in the document]
Comment (amondnet) [reply .please/docs/tracks/active/auto-vercel-build-20260430/plan.md]: ✅ Applied in [f7a132f](https://github.com/amondnet/vercel-action/commit/f7a132f3e4760d11dcdf3cab0e3113e43d907a0f)

- **File**: `plan.md`
- **Change**: Architecture diagram updated to show `[--scope <scope>]` on both commands and `[--output <dir>]` on `vercel build`. Also added `env: VERCEL_TOKEN=<token>, silent: true` annotations to make the auth path explicit.

Implementation already supports all three flags in `src/vercel-build.ts`. Thanks for keeping the docs aligned, @gemini-code-assist!

File: src/vercel-build.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      stdout: (data: Buffer) => {
+        const chunk = data.toString()
+        stdout += chunk
+        core.info(chunk)
+      },
+      stderr: (data: Buffer) => {
+        const chunk = data.toString()
+        stderr += chunk
+        core.info(chunk)
+      },
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/vercel-build.ts:89]: ✅ Applied in [b57e354](https://github.com/amondnet/vercel-action/commit/b57e3546299cd9df80922bd1dd5ef315bab18a54)

- **File**: `src/vercel-build.ts`
- **Change**: Added `appendBounded()` helper that caps captured stdout/stderr at 64 KB per stream (configurable via `MAX_CAPTURED_BYTES`). Live streaming through `core.info()` is unchanged — only the in-memory buffer used to populate `BuildFailedError.stderrTail` is bounded. The trailing 20-line tail behavior is preserved exactly.
- **Test**: New unit test simulates a 1 MB stderr stream and asserts (a) the trailing marker is preserved, (b) the captured buffer stays under 128 KB.

Great catch — long Next.js / monorepo builds could absolutely have hit this. Thanks @copilot-pull-request-reviewer!

File: src/index.ts [hunk 5]
```diff
@@ -299,11 +345,13 @@ async function run(): Promise<void> {
...
+if (!process.env.VITEST) {
+  run().catch((error: unknown) => {
+    if (error instanceof Error) {
+      core.setFailed(error.message)
+    }
+    else {
+      core.setFailed('An unexpected error occurred')
+    }
+  })
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/index.ts]: 💡 Implemented with alternative approach ([b57e354](https://github.com/amondnet/vercel-action/commit/b57e3546299cd9df80922bd1dd5ef315bab18a54))

- **Suggestion**: Use a more specific guard than `!process.env.VITEST` — e.g., `require.main === module`, a dedicated internal env var, or `process.env.GITHUB_ACTIONS === 'true'`.
- **Applied**: Switched to `process.env.GITHUB_ACTIONS === 'true'` (positive guard) — your last option.
- **Reason**: `GITHUB_ACTIONS` is the canonical, runner-set sentinel that's always `'true'` in real workflows and unset/false everywhere else. `require.main === module` doesn't work cleanly because the entry is the bundled `dist/index.js` invoked via `node` and the test imports from source — flipping that semantically is fragile. A custom `__VERCEL_ACTION_TEST` var would also need to be set explicitly in test setup.

Strong catch on the false-positive risk — confirmed locally that `GITHUB_ACTIONS` is unset under vitest, so tests still suppress the auto-invoke. Thanks @copilot-pull-request-reviewer!

File: src/config.ts [hunk 6]
```diff
@@ -85,6 +85,16 @@ export function getActionConfig(): ActionConfig {
...
 
+  const prebuilt = core.getInput('prebuilt') === 'true'
+  const vercelBuild = core.getInput('vercel-build') === 'true'
+  if (prebuilt && vercelBuild) {
+    throw new Error(
+      'Inputs "vercel-build" and "prebuilt" are mutually exclusive. '
+      + 'Set "vercel-build: true" to build inside the action, OR set "prebuilt: true" '
+      + 'to deploy an already-built .vercel/output, but not both.',
+    )
+  }
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/config.ts:96]: ✅ Applied in [b57e354](https://github.com/amondnet/vercel-action/commit/b57e3546299cd9df80922bd1dd5ef315bab18a54)

- **File**: `src/config.ts`
- **Change**: Added a fail-fast mutex check between `vercel-build: true` and a non-empty `vercel-args` (using your suggested error wording), placed right after the `prebuilt` mutex check.
- **Tests**: Two new tests — positive case (throws when both are set) and a whitespace-only edge case (`'   '` is treated as empty so the legacy `vercel-args: ''` default doesn't false-trigger).

Real silent-failure mode — exactly the kind of thing that would only show up in production. Thanks @copilot-pull-request-reviewer!

File: dist/index.js [hunk 7]
```diff
@@ -99687,14 +99781,16 @@ async function run() {
...
 }
-run().catch((error) => {
-    if (error instanceof Error) {
-        core.setFailed(error.message);
-    }
-    else {
-        core.setFailed('An unexpected error occurred');
-    }
-});
+if (!process.env.VITEST) {
```
[the remark this answers is not in the document]
Comment (amondnet) [reply dist/index.js]: ✅ Applied in [b57e354](https://github.com/amondnet/vercel-action/commit/b57e3546299cd9df80922bd1dd5ef315bab18a54)

- **File**: `src/index.ts` (and downstream `dist/index.js`)
- **Change**: Replaced the negative `!process.env.VITEST` guard with the positive `process.env.GITHUB_ACTIONS === 'true'` guard. Same fix discussed on the source-file thread by @copilot-pull-request-reviewer; the bundled output is regenerated.

Confirmed locally that `GITHUB_ACTIONS` is unset under vitest, so tests still skip the auto-invoke. Thanks @cubic-dev-ai!

File: src/config.ts [hunk 8]
```diff
@@ -85,6 +85,16 @@ export function getActionConfig(): ActionConfig {
...
   const githubDeploymentEnvInput = core.getInput('github-deployment-environment')
 
+  const prebuilt = core.getInput('prebuilt') === 'true'
+  const vercelBuild = core.getInput('vercel-build') === 'true'
+  if (prebuilt && vercelBuild) {
+    throw new Error(
+      'Inputs "vercel-build" and "prebuilt" are mutually exclusive. '
+      + 'Set "vercel-build: true" to build inside the action, OR set "prebuilt: true" '
+      + 'to deploy an already-built .vercel/output, but not both.',
+    )
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/config.ts:95]: ✅ Applied in [b57e354](https://github.com/amondnet/vercel-action/commit/b57e3546299cd9df80922bd1dd5ef315bab18a54)

- **File**: `src/config.ts`
- **Change**: Added the suggested fail-fast check rejecting `vercel-build: true` + non-empty `vercel-args`. Same fix as @copilot-pull-request-reviewer's thread on this file; suggested wording adopted with minor wording tweaks for clarity.
- **Tests**: Positive case + whitespace-only edge case.

Thanks @cubic-dev-ai!

File: src/github-comments.ts [hunk 9]
```diff
@@ -117,6 +117,76 @@ export async function createCommentOnCommit(
...
+    ❌ Vercel build failed
+
+    Build for commit \`${sha}\` exited with exit code ${exitCode}.
+
+    \`\`\`
+    ${truncatedTail}
+    \`\`\`
+
+    See the GitHub Actions log for the full output. This pull request is
+    being automatically deployed with [vercel-action](https://github.com/marketplace/actions/vercel-action).
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/github-comments.ts]: ✅ Applied in [b57e354](https://github.com/amondnet/vercel-action/commit/b57e3546299cd9df80922bd1dd5ef315bab18a54)

- **File**: `src/github-comments.ts`
- **Change**: Replaced 'this pull request is being automatically deployed' with 'this workflow run is being automatically deployed' in `buildBuildFailureBody`, so the same body works correctly for both PR and commit comment helpers.

Thanks @cubic-dev-ai!

File: README.md [hunk 10]
```diff
@@ -66,7 +66,8 @@ These inputs use the `@vercel/client` API directly instead of the CLI. They are
 | Name                       | Required | Default   | Description                                                        |
 |----------------------------|:--------:|-----------|--------------------------------------------------------------------|
 | target                     |    No    | `preview` | Deployment target: `production` or `preview`                       |
-| prebuilt                   |    No    | `false`   | Deploy prebuilt output (requires `vercel build` step before)       |
+| prebuilt                   |    No    | `false`   | Deploy prebuilt output (requires a prior `vercel build` step). Mutually exclusive with `vercel-build`. |
+| vercel-build               |    No    | `false`   | Run `vercel pull` + `vercel build` inside the action before deploying, then upload `.vercel/output` via the prebuilt path. Mutually exclusive with `prebuilt`. See ["Build inside the action"](#build-inside-the-action-vercel-build) below. |
```
[the remark this answers is not in the document]
Comment (amondnet) [reply README.md]: ✅ Applied in [b57e354](https://github.com/amondnet/vercel-action/commit/b57e3546299cd9df80922bd1dd5ef315bab18a54)

- **File**: `README.md`
- **Change**: Updated the `vercel-build` row's anchor link from `#build-inside-the-action-vercel-build` to `#method-4---build-inside-the-action-vercel-build` so it correctly resolves to the actual GitHub-rendered slug for `#### Method 4 - Build inside the action (\`vercel-build\`)`.

Thanks @cubic-dev-ai!
